### PR TITLE
fix: prevent sdk cache staleness caused by requestedAt advancement in none responses

### PR DESCRIPTION
**Pull request by cre8ivejp**
Fix https://github.com/bucketeer-io/bucketeer/issues/2461

## Summary

Fixes a bug where server SDKs (Go SDK) could end up with permanently stale
feature flag data that persisted until the process was restarted.

### Root Cause

When the SDK polled `GetFeatureFlags` or `GetSegmentUsers` and nothing had
changed, the server returned a "None" response with `RequestedAt: now.Unix()`.
The SDK stored this value and sent it back on the next poll. Over time,
`RequestedAt` drifted far ahead of the last time the SDK actually received data.

When a flag was eventually updated, the server detected the change (via
`FeatureFlagsId` mismatch) and entered the "Diff" path, which filtered features
using `feature.UpdatedAt > requestedAt - 10s`. Because `requestedAt` had been
artificially advanced by many "None" responses, the updated flag's `UpdatedAt`
fell behind the filter threshold and was silently excluded from the diff.

The SDK then stored the new `FeatureFlagsId` (matching the server) but without
the updated feature data — making it permanently stale, since all subsequent
polls returned "None" (IDs matched).

### Bug timeline

```
T=0 SDK initial sync → "All" response → requestedAt = 0
T=60 SDK polls → no changes → "None" → requestedAt = 60 ← starts drifting
T=120 SDK polls → no changes → "None" → requestedAt = 120
T=125 Flag disabled (auto-ops). UpdatedAt = 125. Redis not yet refreshed.
T=180 SDK polls → Redis stale → "None" → requestedAt = 180 ← drifted past UpdatedAt
T=181 Redis cacher refreshes → ffID changes
T=240 SDK polls → ffID mismatch → "Diff" path:
filter: UpdatedAt (125) > requestedAt - 10s (170) → FALSE
→ flag excluded from diff!
SDK stores new ffID but WITHOUT the updated flag → permanently stale
T=300+ SDK polls → ffID matches → "None" forever. Flag is stale until restart.
```

### Changes

- **"None" responses now return `req.RequestedAt` instead of `now.Unix()`**,
  keeping the SDK's time cursor anchored to the last actual data sync.
- **Removed `secondsForAdjustment` (10s buffer)** from the "Diff" filter in
  both `GetFeatureFlags` and `GetSegmentUsers`. This buffer was a workaround
  for the drift problem and is no longer needed since `RequestedAt` now
  accurately reflects the last data sync time.
- **Changed diff filter from `>` to `>=`** to handle the edge case where a
  feature is updated in the same Unix second as the previous response's
  `RequestedAt`.
- **Replaced `time.Sleep` with `require.Eventually`** in e2e tests for
  reliable cache propagation polling instead of fixed delays.

### Affected APIs

- `GetFeatureFlags` — used by server SDKs for feature flag cache sync
- `GetSegmentUsers` — used by server SDKs for segment user cache sync

`GetEvaluations` / `GetEvaluation` (client SDK APIs) are **not affected** —
client SDKs already preserve `evaluatedAt` correctly by not updating it on
empty responses.

### CODE REVIEW

File: pkg/api/api/api_grpc.go [hunk 1]
```diff
@@ -854,7 +850,7 @@ func (s *grpcGatewayService) GetFeatureFlags(
 			continue
 		}
 		// Check for updated flags
-		if feature.UpdatedAt > adjustedRequestedAt {
+		if feature.UpdatedAt >= req.RequestedAt {
 			updatedFeatures = append(updatedFeatures, feature)
 		}
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/api/api/api_grpc.go]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2462/changes/240edb560393fb4455a09edcc0a32f4349e32525.

File: pkg/api/api/api_grpc_test.go [hunk 2]
```diff
@@ -1810,6 +1889,72 @@ func TestGrpcGetFeatureFlags(t *testing.T) {
...
+			},
+			expected: &gwproto.GetFeatureFlagsResponse{
+				FeatureFlagsId:         singleFeatureID,
+				Features:               []*featureproto.Feature{},
+				ArchivedFeatureFlagIds: make([]string, 0),
+				RequestedAt:            timeNow.Unix(),
+				ForceUpdate:            false,
+			},
+			expectedErr: nil,
+		},
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/api/api/api_grpc_test.go]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2462/commits/9edf5d7c71fb4a14cc41819161598253401b0d3f.

File: pkg/api/api/api_grpc_test.go [hunk 3]
```diff
@@ -1321,11 +1322,83 @@ func TestGrpcGetSegmentUsers(t *testing.T) {
...
+			},
+			expected: &gwproto.GetSegmentUsersResponse{
+				SegmentUsers:      make([]*featureproto.SegmentUsers, 0),
+				DeletedSegmentIds: make([]string, 0),
+				RequestedAt:       timeNow.Unix(),
+				ForceUpdate:       false,
+			},
+			expectedErr:           nil,
+			exactRequestedAtMatch: true,
+		},
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/api/api/api_grpc_test.go]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2462/commits/9edf5d7c71fb4a14cc41819161598253401b0d3f.

File: pkg/api/api/api_grpc.go [hunk 4]
```diff
@@ -1018,25 +1017,33 @@ func (s *grpcGatewayService) GetSegmentUsers(
...
 	}
 
+	// Clamp requestedAt to prevent future timestamps (e.g. from cross-server clock skew)
+	// from permanently excluding all updates in the Diff filter.
+	requestedAt := min(req.RequestedAt, time.Now().Unix())
 	// Filter the updated segments.
```
[the remark this answers is not in the document]
Comment (cre8ivejp) [reply pkg/api/api/api_grpc.go]: Fixed https://github.com/bucketeer-io/bucketeer/pull/2462/commits/9edf5d7c71fb4a14cc41819161598253401b0d3f.
